### ISSUE TITLE
for-each statement is faster than `forEach` method

### ISSUE THREAD

**Issue by scheglov** [dart-lang/sdk#53817]
See https://dart-review.googlesource.com/c/sdk/+/330983 for context.

It is probably known, and this is why we have the corresponding lint.

**Comment by lrhn** [comment]
It's fairly well known. The lack of overheard of a method call, probably polymorphic if `forEach` is not inlined, makes `for`-`in` usually be faster.

It's not guaranteed. If the loop compilation doesn't recognize the Iterable, and cannot inline the iterator, then it's possible, at least hypothetically, that the `forEach` can do the iteration more efficiently.

**Comment by scheglov** [comment]
Yeah, I see that the performance is quite different in different cases.

Specifically this code is not important to me at all, I just wanted to see what happens, how much it is not guarantee.
There are surprises.
At least for-each statement is reliably 2x slower in AOT :-)

<details>
  <summary>Code is here</summary>
  
```dart
// ignore: unused_import
import 'dart:typed_data';

void main() {
  const repeatCount = 1000;
  final values = Int64List(1000000);
  // final values = List.filled(1000000, 0);
  {
    final timer = Stopwatch()..start();
    for (var i = 0; i < repeatCount; i++) {
      doForEachStatement(values);
    }
    print('doForEachStatement: ${timer.elapsedMilliseconds} ms');
  }
  {
    final timer = Stopwatch()..start();
    for (var i = 0; i < repeatCount; i++) {
      doForEachMethod(values);
    }
    print('doForEachMethod: ${timer.elapsedMilliseconds} ms');
  }
}

int doForEachMethod(List<int> values) {
  var result = 0;
  values.forEach((value) {
    result = (result + value) & 0xFFFF;
  });
  return result;
}

int doForEachStatement(List<int> values) {
  var result = 0;
  for (final value in values) {
    result = (result + value) & 0xFFFF;
  }
  return result;
}
```
</details>

JIT, Int64List, doForEachStatement: 1212 ms
JIT, Int64List, doForEachMethod: 4791 ms

AOT, Int64List, doForEachStatement: 2448 ms
AOT, Int64List, doForEachMethod: 2999 ms

JIT, List\<int\>, doForEachStatement: 649 ms
JIT, List\<int\>, doForEachMethod: 4884 ms

AOT, List\<int\>, doForEachStatement: 1266 ms
AOT, List\<int\>, doForEachMethod: 390 ms

**Comment by lrhn** [comment]
For `forEach`, it's necessary to say _which_ `forEach` it is. Each `Iterable` can have its own `forEach`, with widely different implementation and performance. 
That's one reason it's harder to optimize, you don't always know which function is being called. And when you do, unless it's a platform type, the best the compiler can do is to inline.
For platform types, it's possible to optimize the iteration just as much as `for`-`in`. It's probably just not done.

**Comment by mraleph** [comment]
> AOT, List<int>, doForEachStatement: 1266 ms
> AOT, List<int>, doForEachMethod: 390 ms

In AOT both of these are fully inlined into `main` and specialised, but in `for-in` benchmark we have the following piece left over from the list iterator implementation:

```
148:     v325 <- LoadIndexed(v84, v287 T{int}) T{*?}
...
152:     Branch if StrictCompare:16(===, v325 T{X0??}, v0 T{Null?}) goto (34, 35)
154: B34[target]:20
...
156:     t0 <- AssertAssignable:22(v325 T{X0??}, v290, ' in type cast', instantiator_type_args(v0 T{Null}), function_type_args(v0 T{Null?})) [-9223372036854775808, 9223372036854775807] T{int}
158:     ParallelMove  goto:28 B36
160: B35[target]:24
162:     ParallelMove  goto:30 B36
164: B36[join]:26 pred(B34, B35) ParallelMove rcx <- C
```

These are remnants of `_current as E` - transformed into `_current == null ? _current as E : _current`.

Compiler failed to eliminate them because it failed to figure out the type of `LoadIndexed(...)`. This in turn happened because `CreateArray` instruction is missing proper type arguments inference rule. Changing it like this fixes it: 

```cpp
CompileType CreateArrayInstr::ComputeType() const {
  if (type_arguments()->BindsToConstant() &&
      type_arguments()->BoundConstant().IsTypeArguments()) {
    const auto& type_args =
        TypeArguments::Cast(type_arguments()->BoundConstant());
    return CompileType(
        /*can_be_null=*/false, /*can_be_sentinel=*/false, kArrayCid,
        &AbstractType::ZoneHandle(Type::New(
            Class::Handle(
                IsolateGroup::Current()->object_store()->array_class()),
            type_args, Nullability::kNonNullable)));
  }
  return CompileType::FromCid(kArrayCid);
}
```

After this change both loop become the same, however on my box we still see a performance difference:

```
doForEachMethod: 558 ms
doForEachStatement: 861 ms
```

Note that both loops are empty at this point (DCE eliminated loading from array and arithmetic operations because their results are never used). 

The code placement (loop alignment) is to blame for the remaining difference between two empty loops. If we align all loop headers in the program the difference disappears: 

```
doForEachMethod: 560 ms
doForEachStatement: 551 ms
```

In `Int64List` compiler does not inline `_TypedListBase.get iterator`. Putting `@pragma('vm:prefer-inline')` on `get iterator` gets us to:

```
doForEachMethod: 3417 ms
doForEachStatement: 552 ms
```

Final remark: it'd caution against making choices between `forEach` and `for-in` if your loop is making any significant amount of work. If it is not just summing elements of the array - the difference between `forEach` and `for-in` might not matter.

**Comment by scheglov** [comment]
Thank you for the analysis, it is always so enlightening to me.

Yeah, I know that benchmark results should be viewed with some caution.
FWIW, it was motivated by a [change](https://dart-review.googlesource.com/c/sdk/+/330983) in the actual code where switching to `for-in` produced nice performance benefits.

**Comment by mkustermann** [comment]
> FWIW, it was motivated by a [change](https://dart-review.googlesource.com/c/sdk/+/330983) in the actual code where switching to for-in produced nice performance benefits.

The AOT can do a good job in this case, because it's easy to infer the iterable to be always an `IntSet`. In cases where we cannot infer the concrete type a `.forEach()` may often be faster in AOT (as 2 non-inlinable `.moveNext()` / `.current` calls may be slower than a single closure call) - but of course if one does a lot of work per-element the overhead of iteration doesn't matter.

**Comment by scheglov** [comment]
In the benchmark below `forEach()` is always slower than `for-in`.

JIT, always-inline, doForEachStatement, Int64List: 1234 ms
JIT, always-inline, doForEachStatement, Int32List: 996 ms
JIT, always-inline, doForEachMethod, Int64List: 5001 ms
JIT, always-inline, doForEachMethod, Int32List: 4968 ms

AOT, always-inline, doForEachStatement, Int64List: 1377 ms
AOT, always-inline, doForEachStatement, Int32List: 2614 ms
AOT, always-inline, doForEachMethod, Int64List: 3143 ms
AOT, always-inline, doForEachMethod, Int32List: 3136 ms

JIT, never-inline, doForEachStatement, Int64List: 1230 ms
JIT, never-inline, doForEachStatement, Int32List: 991 ms
JIT, never-inline, doForEachMethod, Int64List: 4990 ms
JIT, never-inline, doForEachMethod, Int32List: 4988 ms

AOT, never-inline, doForEachStatement, Int64List: 2611 ms
AOT, never-inline, doForEachStatement, Int32List: 2616 ms
AOT, never-inline, doForEachMethod, Int64List: 3142 ms
AOT, never-inline, doForEachMethod, Int32List: 3151 ms

And I'm really glad that AOT can do good job inlining, in my outside opinion all variants of writing iterations: `forEach()`, `for-in` and `for-index` - if inlinable, over a `List` with known length, should produce the identical code.

<details>
  <summary>Code is here</summary>
  
```dart
// ignore: unused_import
import 'dart:typed_data';

void main() {
  const repeatCount = 1000;
  final values1 = Int64List(1000000);
  final values2 = Int32List(1000000);
  int result = 0;
  {
    final timer = Stopwatch()..start();
    for (var i = 0; i < repeatCount; i++) {
      result = (result + doForEachStatement(values1)) & 0xFF;
    }
    print('doForEachStatement, Int64List: ${timer.elapsedMilliseconds} ms');
  }
  {
    final timer = Stopwatch()..start();
    for (var i = 0; i < repeatCount; i++) {
      result = (result + doForEachStatement(values2)) & 0xFF;
    }
    print('doForEachStatement, Int32List: ${timer.elapsedMilliseconds} ms');
  }
  {
    final timer = Stopwatch()..start();
    for (var i = 0; i < repeatCount; i++) {
      result = (result + doForEachMethod(values1)) & 0xFF;
    }
    print('doForEachMethod, Int64List: ${timer.elapsedMilliseconds} ms');
  }
  {
    final timer = Stopwatch()..start();
    for (var i = 0; i < repeatCount; i++) {
      result = (result + doForEachMethod(values2)) & 0xFF;
    }
    print('doForEachMethod, Int32List: ${timer.elapsedMilliseconds} ms');
  }
  print('result: $result');
}

@pragma('vm:always-inline')
// @pragma('vm:never-inline')
int doForEachMethod(List<int> values) {
  var result = 0;
  values.forEach((value) {
    result = (result + value) & 0xFFFF;
  });
  return result;
}

@pragma('vm:always-inline')
// @pragma('vm:never-inline')
int doForEachStatement(List<int> values) {
  var result = 0;
  for (final value in values) {
    result = (result + value) & 0xFFFF;
  }
  return result;
}
```
</details>
